### ISSUE TITLE
Speed issue for TLL family

### ISSUE THREAD

**Issue by tvatter** [vinecopulib/vinecopulib#355]
See https://github.com/vinecopulib/rvinecopulib/issues/121

**Comment by tvatter** [comment]
I'm using our test suit (with 2000 observations instead of 100) with a timer class and the following

```
Timer timer;

    timer.start();
    bicop_.fit(u, controls);
    std::cout << "fit: " << timer.elapsedMilliseconds() << std::endl;
    timer.start();

    EXPECT_GE(bicop_.pdf(u).minCoeff(), 0.0);
    std::cout << "pdf: " << timer.elapsedMilliseconds() << std::endl;
    timer.start();

    EXPECT_GE(bicop_.hfunc1(u).minCoeff(), 0.0);
    std::cout << "hfunc1: " << timer.elapsedMilliseconds() << std::endl;
    timer.start();

    EXPECT_GE(bicop_.hfunc2(u).minCoeff(), 0.0);
    std::cout << "hfunc2: " << timer.elapsedMilliseconds() << std::endl;
    timer.start();

    EXPECT_GE(bicop_.hinv1(u).minCoeff(), 0.0);
    std::cout << "hinv1: " << timer.elapsedMilliseconds() << std::endl;
    timer.start();

    EXPECT_GE(bicop_.hinv2(u).minCoeff(), 0.0);
    std::cout << "hinv2: " << timer.elapsedMilliseconds() << std::endl;
```
Result:
<img width="535" alt="screen shot 2018-06-14 at 17 16 51" src="https://user-images.githubusercontent.com/5273142/41438868-c8fadaac-6ff6-11e8-9977-e6c9989cced4.png">

**Comment by tvatter** [comment]
Using the stuff below (i.e., stupid integration), we are 25% faster for hfuncs/hinvs.

```
inline double InterpolationGrid::int_on_grid(const double &upr,
                                             const Eigen::VectorXd &vals,
                                             const Eigen::VectorXd &grid)
{
    ptrdiff_t m = grid.size();
    double newint;

    double tmpint = 0.0;

    if (upr > grid(0)) {
        // go up the grid and integrate
        for (ptrdiff_t k = 0; k < m - 1; ++k) {
            // stop loop if fully integrated
            if (upr < grid(k))
                break;

            // don't integrate over full cell if upr is in interior
            if (upr > grid(k)) {
                tmpint += vals(k) * (upr - grid(k));
            } else {
                tmpint += (vals(k + 1) + vals(k)) * (grid(k + 1) - grid(k)) / 2.0;
            }
        }
    }

    return tmpint;
}
```

**Comment by tvatter** [comment]
Second suggestion, use bilinear bilinear interpolation:
```
inline double
bilinear_interpolation(double z11,
                       double z12,
                       double z21,
                       double z22,
                       double x1,
                       double x2,
                       double y1,
                       double y2,
                       double x,
                       double y)
{
    double x2x1, y2y1, x2x, y2y, yy1, xx1;
    x2x1 = x2 - x1;
    y2y1 = y2 - y1;
    x2x = x2 - x;
    y2y = y2 - y;
    yy1 = y - y1;
    xx1 = x - x1;
    return (z11 * x2x * y2y +
            z21 * xx1 * y2y +
            z12 * x2x * yy1 +
            z22 * xx1 * yy1) / (x2x1 * y2y1);
}

//! Interpolation in two dimensions
//!
//! @param x mx2 matrix of evaluation points.
inline Eigen::VectorXd
InterpolationGrid::interpolate(const Eigen::MatrixXd &x)
{

    ptrdiff_t m = grid_points_.size();
    auto f = [&m, this](double x0, double x1) {

        Eigen::Matrix<ptrdiff_t, 1, 2> ij = this->get_ij(x0, x1, m);
        return fmax(bilinear_interpolation(
            this->values_(ij(0), ij(1)),
            this->values_(ij(0), ij(1) + 1),
            this->values_(ij(0) + 1, ij(1)),
            this->values_(ij(0) + 1, ij(1) + 1),
            this->grid_points_(ij(0)),
            this->grid_points_(ij(0) + 1),
            this->grid_points_(ij(1)),
            this->grid_points_(ij(1) + 1),
            x0,
            x1), 1e-15);
    };

    return tools_eigen::binaryExpr_or_nan(x, f);
}
```
With this, we have:
<img width="592" alt="screen shot 2018-06-14 at 18 32 03" src="https://user-images.githubusercontent.com/5273142/41441582-462933fc-7001-11e8-96b3-419840c626d7.png">

**Comment by tnagler** [comment]
solved by #357
